helix-term/src/ui/statusline.rs

```diff
@@ -0,0 +1,349 @@
+use helix_core::{coords_at_pos, encoding};
+use helix_view::{
+    document::{Mode, SCRATCH_BUFFER_NAME},
+    graphics::Rect,
+    theme::Style,
+    Document, Editor, Theme, View,
+};
+
+use crate::ui::ProgressSpinners;
+
+use helix_view::editor::StatusLineElement as StatusLineElementID;
+use tui::buffer::Buffer as Surface;
+use tui::text::{Span, Spans};
+
+/// A status line element contains the information about a component which can be displayed in the status line.
+struct StatusLineElement {
+    /// The element
+    pub text: String,
+
+    /// The style to be used to render the element (this style will be merged with the base style).
+    /// If not set, a default base style will be used.
+    pub style: Option<Style>,
+}
```

Comment:
Is this extra struct really needed? You could map the config enum directly to simple functions that take a render context. Gutters work in a similar way but are a bit more complicated (a generator function that returns a closure since we repeatedly call it in a single render)

Reply:
The other advantage of exposing functions directly: you could allocate a single buffer for each section, then pass it into the render call and use `write!` instead of `format!`. That way you're not allocating a new string for each part.

i.e. for gutters

https://github.com/helix-editor/helix/blob/fcb8c50c9800338966b015863887f9f1c4f62659/helix-term/src/ui/editor.rs#L560-L561=

then each component just writes into it:

https://github.com/helix-editor/helix/blob/fcb8c50c9800338966b015863887f9f1c4f62659/helix-view/src/gutter.rs#L68=

Reply:
Thanks again for the review, @archseer !

I finally found the time to have a look at it and got rid of `StatusLineElement` (see 5fe398d).
I took a more functional approach and simplified/generalized `get_render_function()` so it directly maps from an enum to a function. I like it much more like this; I hope that's more or less how you imagined it.

I didn't get rid of `format!()` in favor of `write!()` just yet, as it would require me obtaining an output stream directly instead of working with `Spans` buffers, which it turn would probably require refactoring beyond **statusline.rs**; I think that would become a bit too much for this PR.